core/runtime/src/main/java/io/quarkus/runtime/configuration/SimpleConfigurationProviderResolver.java

```diff
@@ -10,14 +12,11 @@
  * A simple configuration provider.
  */
 public class SimpleConfigurationProviderResolver extends ConfigProviderResolver {
-    private final Config config;
 
-    public SimpleConfigurationProviderResolver(final Config config) {
-        this.config = config;
-    }
+    private static final AtomicReference<Config> CONFIG = new AtomicReference<>();
```

Comment:
The usage follows from the "incorrect" usage of static fields in MP Config API. Even if we replace the `ConfigurationProviderResolver` the `ConfigProvider` will continue using the old instance. I prefer using `AtomicReference` over `volatile`.

Reply:
Efficient systems are made from many small optimizations, as our friend Andrew Haley likes to say, but I would also say that inefficient systems are made from many small deoptimizations.  There's really no good justification for this preference; I would advise using a volatile in all cases as it is 4 or 8 bytes versus 68 or more bytes plus an allocation plus an extra object on the heap forever.

Reply:
Can you explain the MP config behavior you're attempting to rectify?

Reply:
Ok, I'll try once again. We currently invoke  `org.eclipse.microprofile.config.spi.ConfigProviderResolver.setInstance(ConfigProviderResolver)` in `RUNTIME_INIT` when an app starts. But `org.eclipse.microprofile.config.ConfigProvider` only obtains the resolver once when initializing `org.eclipse.microprofile.config.ConfigProvider.INSTANCE`. As a result, if you run two QuarkusUnitTests and both inject/obtain config properties and have different config sources, the first one wins and the `Config` created for the second one is just ignored.

Reply:
OK I see.  This solution is probably OK for the short term, but the best long term solution is to either go back to the `SmallRyeConfigProviderResolver` or provide a similar impl (maybe one which gives an error when no config is loaded) where we can reset the config.  I have this going in a branch but there are some problems relating to class loading.  So that part will have to wait until after Alexey's change comes in.

Reply:
Yes, it's a short term hack. I'll try to address your objections (even though I can't say I agree).

> 4 or 8 bytes versus 68 plus an extra object on the heap 

Sounds correct but with this approach we would have to refactor most of all code ;-)